File: clients/apple/Sources/PunktfunkKit/Gamepad/DeviceGyro.swift
// The opt-in phone-gyro mirror (`DefaultsKey.gyroFromDevice`): when player 1's forwarded
// controller has no rotation sensor of its own, THIS device's IMU speaks for it on the wire's
// motion plane — for clip-on and third-party pads that ship without a gyro, where the phone
// body is rigidly attached to (or simply is) the thing in the player's hands. The sibling of
// `GamepadFeedback`'s rumble-on-device mirror, with the data flowing the other way.
//
// GamepadCapture owns the engage/stand-down decision (it knows the pad-0 slot and whether its
// controller reports a rotation rate); this class only turns CoreMotion on and off and converts
// samples. Two invariants it enforces itself:
//  - one motion writer per pad: samples go out only between `start` and `stop`, and capture
//    suppresses pad 0's controller-motion forwarding while this runs;
//  - no stale rotation: `stop` sends a single zero-gyro sample after the last real one, so the
//    host's virtual pad never keeps integrating an angular velocity this device stopped
//    producing (the gyro-sweep "stale angular velocity re-sent forever" failure mode).
//
// Samples are CMDeviceMotion (sensor-fused: bias-corrected rotation rate, gravity split from
// user acceleration) at the ~100 Hz CoreMotion ceiling — below a DualSense's 250 Hz, but the
// host's motion plane is event-driven, not cadence-locked, so a slower producer just means
// fewer samples. Units and axis semantics match `GamepadCapture.forwardMotion` exactly (the
// `GamepadWire` constants; accel = gravity + user acceleration — the same convention, so a
// future sign/scale correction lands in one place for both sources). The one thing the phone
// adds is a frame remap: CoreMotion reports in the device's portrait frame, while the wire
// wants the controller frame the player sees (x right, y up, z out of the screen), so each
// sample is rotated by the current interface orientation — a phone clipped landscape must yaw
// when the player yaws, not roll.

#if os(iOS)
import CoreMotion
import Foundation
import UIKit

/// Device-frame → controller-frame axis remap for one interface orientation. CoreMotion's
/// frame is fixed to the portrait device (+x right edge, +y top, +z out of the screen); the
/// controller frame keeps +z (the screen always faces the player) and rotates x/y so they
/// mean "player's right" and "player's up". Derived, like the wire scale constants — pinned
/// by `DeviceGyroRemapTests`, correctable in one place if on-glass says otherwise.
/// File-scope rather than nested so the sample thread can use it without actor isolation.
enum DeviceGyroRemap {
    case identity
    /// Upside-down portrait: both in-plane axes flip.
    case flipped
    /// Landscape, device top to the player's LEFT (interface `.landscapeRight`):
    /// player-right = device-bottom, player-up = device-right.
    case topLeft
    /// Landscape, device top to the player's RIGHT (interface `.landscapeLeft`).
    case topRight

    init(_ orientation: UIInterfaceOrientation) {
        switch orientation {
        case .portraitUpsideDown: self = .flipped
        case .landscapeRight: self = .topLeft
        case .landscapeLeft: self = .topRight
        default: self = .identity
        }
    }

    /// Rotate one device-frame vector (rotation rate or acceleration — both transform the
    /// same way under an in-plane rotation) into the controller frame.
    func apply(x: Float, y: Float, z: Float) -> (x: Float, y: Float, z: Float) {
        switch self {
        case .identity: return (x, y, z)
        case .flipped: return (-x, -y, z)
        case .topLeft: return (-y, x, z)
        case .topRight: return (y, -x, z)
        }
    }
}

@MainActor
public final class DeviceGyro {
    /// Whether this device can source motion at all — gates the settings rows (a device
    /// without an IMU would make the toggle a silent no-op, the rumble mirror's rule).
    /// One shared probe: Apple recommends a single `CMMotionManager` per app, and the
    /// settings UI asking per-render must not allocate one each time.
    public static let isAvailable: Bool = CMMotionManager().isDeviceMotionAvailable

    /// Everything the sample thread touches, behind one lock: the orientation remap (written
    /// on main when the device rotates), the last converted accel, and whether a real sample
    /// went out (so `stop` knows it owes the wire a zero). Kept off the actor deliberately —
    /// `forward` runs on the delivery queue.
    private final class SampleState: @unchecked Sendable {
        let lock = NSLock()
        var remap: DeviceGyroRemap = .identity
        var sentSample = false
        /// Re-sent with the closing zero-gyro sample so "rotation stopped" doesn't also
        /// overwrite a plausible gravity vector with free-fall.
        var lastAccel: (Int16, Int16, Int16) = (0, 0, 0)
    }

    /// Ship one converted sample (wire pad 0). Must be thread-safe — invoked from the
    /// delivery queue (`PunktfunkConnection.sendMotion` locks internally).
    private let send: @Sendable (_ gyro: (Int16, Int16, Int16), _ accel: (Int16, Int16, Int16)) -> Void

    private let motion = CMMotionManager()
    /// Dedicated serial delivery queue — deliberately NOT main (the controller path's
    /// main-queue delivery is a known jitter source; the mirror starts clean).
    private let queue: OperationQueue = {
        let q = OperationQueue()
        q.name = "punktfunk.device-gyro"
        q.maxConcurrentOperationCount = 1
        return q
    }()

    private let state = SampleState()
    private var orientationObserver: NSObjectProtocol?

    /// Whether the mirror is between `start` and `stop` — read by GamepadCapture to keep the
    /// controller path off pad 0's motion while this runs.
    public private(set) var isRunning = false

    public init(
        send: @escaping @Sendable (_ gyro: (Int16, Int16, Int16), _ accel: (Int16, Int16, Int16)) -> Void
    ) {
        self.send = send
    }

    /// Begin sourcing pad-0 motion from this device. Idempotent.
    public func start() {
        guard !isRunning, motion.isDeviceMotionAvailable else { return }
        isRunning = true
        updateRemap()
        // Interface orientation only changes alongside a device-orientation notification, so
        // this is the one signal needed; re-reading the scene keeps a rotation lock stable.
        orientationObserver = NotificationCenter.default.addObserver(
            forName: UIDevice.orientationDidChangeNotification, object: nil, queue: .main
        ) { [weak self] _ in
            MainActor.assumeIsolated { self?.updateRemap() }
        }
        // CoreMotion's practical ceiling; requesting faster just clamps.
        motion.deviceMotionUpdateInterval = 1.0 / 100.0
        motion.startDeviceMotionUpdates(to: queue) { [state, send] m, _ in
            guard let m else { return }
            Self.forward(m, state: state, send: send)
        }
    }

    /// Stop sourcing and, if anything was sent, park the host pad's rotation at zero. The
    /// zero rides the same serial queue as the samples, so it is guaranteed last — without
    /// blocking the caller.
    public func stop() {
        guard isRunning else { return }
        isRunning = false
        motion.stopDeviceMotionUpdates()
        if let o = orientationObserver {
            NotificationCenter.default.removeObserver(o)
            orientationObserver = nil
        }
        queue.addOperation { [state, send] in
            state.lock.lock()
            let owed = state.sentSample
            state.sentSample = false
            let accel = state.lastAccel
            state.lock.unlock()
            if owed { send((0, 0, 0), accel) }
        }
    }

    private func updateRemap() {
        let o = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first?.interfaceOrientation ?? .portrait
        state.lock.lock()
        state.remap = DeviceGyroRemap(o)
        state.lock.unlock()
    }

    /// Runs on the delivery queue: remap, scale, ship.
    nonisolated private static func forward(
        _ m: CMDeviceMotion, state: SampleState,
        send: (_ gyro: (Int16, Int16, Int16), _ accel: (Int16, Int16, Int16)) -> Void
    ) {
        state.lock.lock()
        let r = state.remap
        state.lock.unlock()
        let rot = r.apply(
            x: Float(m.rotationRate.x), y: Float(m.rotationRate.y), z: Float(m.rotationRate.z))
        // Total acceleration, NEGATED — the same convention as GamepadCapture.forwardMotion, which
        // this file's header promises to track. Apple reports the gravity VECTOR (pointing down);
        // an accelerometer measures proper acceleration (pointing up at rest), and the wire carries
        // the latter. Without the minus a still phone told the host it was accelerating downward at
        // 1 g.
        let acc = r.apply(
            x: -Float(m.gravity.x + m.userAcceleration.x),
            y: -Float(m.gravity.y + m.userAcceleration.y),
            z: -Float(m.gravity.z + m.userAcceleration.z))
        // NO frame conversion here, and that is not an oversight — `GamepadCapture.forwardMotion`
        // applies `GamepadWire.appleMotionToWire` and this deliberately does not.
        //
        // The trap is that two different frames are both called "the controller frame". GCMotion
        // reports a CONTROLLER in (Right, Forward, Up) — measured on a real DualSense — which is
        // not the wire's frame, hence the conversion over there. `r` above resolves THIS DEVICE
        // into the frame the header describes: x right, y up, z out of the screen. For the pose
        // this mirror exists to serve — a phone clipped upright, screen facing the player — "out of
        // the screen" points AT the player, so that frame is (Right, Up, Backward), which IS the
        // wire's frame. Straight through is already correct.
        //
        // Applying the controller path's conversion here was tried and was WRONG: a phone at rest
        // would have reported gravity as −1 g on the roll axis instead of +1 g up, i.e. lying on
        // its edge. Caught by measuring the Android twin, which does the same thing straight
        // through and reads +1 g on the up axis end to end. If a future capture path needs a
        // conversion, decide it from that source's OWN measured frame rather than by analogy.
        let gs = GamepadWire.gyroLSBPerRadS
        let as_ = GamepadWire.accelLSBPerG
        let gyro = (
            GamepadWire.motionRaw(rot.x, scale: gs),
            GamepadWire.motionRaw(rot.y, scale: gs),
            GamepadWire.motionRaw(rot.z, scale: gs)
        )
        let accel = (
            GamepadWire.motionRaw(acc.x, scale: as_),
            GamepadWire.motionRaw(acc.y, scale: as_),
            GamepadWire.motionRaw(acc.z, scale: as_)
        )
        state.lock.lock()
        state.lastAccel = accel
        state.sentSample = true
        state.lock.unlock()
        send(gyro, accel)
    }
}
#endif

File: clients/android/kit/src/main/kotlin/io/unom/punktfunk/kit/DeviceGyro.kt
package io.unom.punktfunk.kit

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.view.Display
import android.view.Surface
import android.view.WindowManager

/**
 * The opt-in phone-gyro mirror ("Gyro from this phone", off by default): while wire pad 0 is a
 * controller with no motion source of its own, THIS device's IMU speaks for it on the rich-input
 * motion plane — for clip-on and third-party pads that ship without a gyro, where the phone body
 * is rigidly attached to (or simply is) the thing in the player's hands. [GamepadFeedback]'s
 * rumble-on-phone mirror with the data flowing the other way.
 *
 * On Android the only motion sources are the capture links (USB DualSense / SC2 — pads with a
 * real IMU, claimed as [GamepadRouter.ExternalPad]s), so the stand-down rule is exactly
 * [GamepadRouter.padHasOwnMotion]: when a capture link holds pad 0, the mirror sends nothing —
 * two motion writers on one wire pad would fight. It also sends nothing while pad 0 has no slot
 * at all (motion never creates a host pad; a controller must have arrived first).
 *
 * Two properties this class enforces itself:
 *  - samples ride a dedicated [HandlerThread] with batching disabled (`maxReportLatencyUs = 0`) —
 *    sensor batching would trade the exact latency gyro aim exists to avoid;
 *  - a stand-down edge (capture link claims pad 0, or [stop]) sends ONE zero-gyro sample, so the
 *    host's virtual pad never keeps integrating an angular velocity this device stopped
 *    producing (the gyro-sweep "stale angular velocity re-sent forever" failure mode).
 *
 * Units are the wire contract, converted by [Gamepad.motionGyroWire] / [Gamepad.motionAccelWire] —
 * the same two functions [PadSensors] uses, so a scale this client ever has to correct is corrected
 * once for every sender rather than once per sender that someone remembers. The one thing the
 * phone adds is a frame remap: sensors report in the device's natural-portrait frame, while
 * the wire wants the controller frame the player sees (x right, y up, z out of the screen), so
 * each sample is rotated by the current display rotation — a phone clipped landscape must yaw
 * when the player yaws, not roll. The matrix is derived and pinned by `DeviceGyroTest`;
 * correctable in one place if on-glass says otherwise.
 */
class DeviceGyro(
    context: Context,
    private val handle: Long,
    private val router: GamepadRouter,
) : SensorEventListener {

    private val sensorManager: SensorManager? =
        context.getSystemService(SensorManager::class.java)

    /** For the live rotation; null on contexts without a display association (then portrait). */
    private val display: Display? = runCatching {
        if (Build.VERSION.SDK_INT >= 30) {
            context.display
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(WindowManager::class.java)?.defaultDisplay
        }
    }.getOrNull()

    private val thread = HandlerThread("pf-phone-gyro")

    /** Latest converted accel, paired with each gyro send (the wire fuses both per sample). */
    private val lastAccel = intArrayOf(0, Gamepad.MOTION_ACCEL_LSB_PER_G, 0)

    /** Whether the last gyro event actually went to pad 0 — the stand-down zero-send edge. */
    private var wasWriting = false

    /** Register the listeners; a device without a gyroscope makes this a no-op. */
    fun start() {
        val sm = sensorManager ?: return
        val gyro = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE) ?: return
        thread.start()
        val h = Handler(thread.looper)
        // ~200 Hz requested (the framework clamps to what the hardware offers), zero report
        // latency: batching is poison for gyro aim.
        sm.registerListener(this, gyro, SAMPLING_PERIOD_US, 0, h)
        sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)?.let {
            sm.registerListener(this, it, SAMPLING_PERIOD_US, 0, h)
        }
    }

    /**
     * Unregister and join the sensor thread, then park the host pad's rotation at zero if this
     * mirror was the live writer. Call BEFORE the router is released / the handle freed —
     * teardown-ordered like the feedback threads.
     */
    fun stop() {
        sensorManager?.unregisterListener(this)
        thread.quitSafely()
        runCatching { thread.join() }
        if (wasWriting) {
            wasWriting = false
            sendZero()
        }
    }

    override fun onSensorChanged(event: SensorEvent) {
        val rotation = display?.rotation ?: Surface.ROTATION_0
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                val v = remap(rotation, event.values[0], event.values[1], event.values[2])
                for (i in 0..2) lastAccel[i] = Gamepad.motionAccelWire(v[i])
            }
            Sensor.TYPE_GYROSCOPE -> {
                // The write gate, per sample: sends must be on at all (the forwarding
                // preference AND the session's GAMEPAD grant — an AccessUpdate can revoke it
                // mid-session), pad 0 must exist (motion never creates a pad) and must not be
                // a capture link's (its own IMU is streaming).
                val write = router.sendsEnabled() && router.padPresent(0) &&
                    !router.padHasOwnMotion(0)
                if (!write) {
                    // Stand-down edge: never leave the last angular velocity latched host-side.
                    if (wasWriting) {
                        wasWriting = false
                        sendZero()
                    }
                    return
                }
                wasWriting = true
                val v = remap(rotation, event.values[0], event.values[1], event.values[2])
                NativeBridge.nativeSendPadMotion(
                    handle, 0,
                    Gamepad.motionGyroWire(v[0]),
                    Gamepad.motionGyroWire(v[1]),
                    Gamepad.motionGyroWire(v[2]),
                    lastAccel[0], lastAccel[1], lastAccel[2],
                )
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

    /** Zero rotation, last-known accel — "at rest", not free-fall. */
    private fun sendZero() {
        NativeBridge.nativeSendPadMotion(
            handle, 0, 0, 0, 0, lastAccel[0], lastAccel[1], lastAccel[2],
        )
    }

    companion object {
        /** Whether this device can source motion at all — gates the settings rows (a TV box
         *  without an IMU would make the toggle a silent no-op, the rumble mirror's rule). */
        fun available(context: Context): Boolean =
            context.getSystemService(SensorManager::class.java)
                ?.getDefaultSensor(Sensor.TYPE_GYROSCOPE) != null

        /**
         * ~200 Hz — between the sensor's usual FASTEST (~250-500 Hz) and GAME (~50 Hz), and also
         * the ceiling the framework grants an app without `HIGH_SAMPLING_RATE_SENSORS` (API 31+),
         * so asking for more would only be silently capped. Shared with [PadSensors].
         */
        internal const val SAMPLING_PERIOD_US = 5000

        /**
         * Rotate one device-frame vector (rotation rate or acceleration — both transform the
         * same way under an in-plane rotation) into the controller frame for [rotation]
         * ([Surface].ROTATION_*). Sensors report in the natural-portrait frame (+x right edge,
         * +y top, +z out of the screen); the controller frame keeps +z (the screen always faces
         * the player) and rotates x/y to mean "player's right" and "player's up". ROTATION_90 =
         * the device physically turned counter-clockwise, top to the player's LEFT.
         */
        fun remap(rotation: Int, x: Float, y: Float, z: Float): FloatArray = when (rotation) {
            Surface.ROTATION_90 -> floatArrayOf(-y, x, z) // top left: right = bottom, up = +x
            Surface.ROTATION_270 -> floatArrayOf(y, -x, z) // top right: right = top, up = −x
            Surface.ROTATION_180 -> floatArrayOf(-x, -y, z)
            else -> floatArrayOf(x, y, z)
        }
    }
}
